### PR TITLE
Events buffer

### PR DESCRIPTION
**Pull request by sadym-chromium**
Implementation of the [log events buffering](https://w3c.github.io/webdriver-bidi/#buffer-a-log-event).
* Added `IdWrapper` abstraction.
* Added `Buffer` abstraction.
* Store buffers and "last send event per event name, channel and browsing context".

Not implemented yet:
* e2e testing.

### CODE REVIEW

File: src/bidiMapper/domains/events/EventManager.ts [hunk 1]
```diff
@@ -86,4 +144,89 @@ export class EventManager implements IEventManager {
...
+      Math.max(this.#lastMessageSent.get(lastSentMapKey) ?? 0, eventWrapper.id)
+    );
+  }
+
+  /**
+   * Returns events which are buffered and not yet sent to the given channel events.
+   * @param eventName
+   * @param contextId
+   * @param channel
+   * @private
```
Comment (thiagowfx) [inline src/bidiMapper/domains/events/EventManager.ts:209]: What is the point of the `@private` annotation, given that the function is marked with `#` already?

File: src/bidiMapper/domains/events/EventManager.ts [hunk 2]
```diff
@@ -47,30 +71,64 @@ export class EventManager implements IEventManager {
     this.#subscriptionManager = new SubscriptionManager();
   }
 
+  /**
+   * Returns consistent key to be used to access values maps.
+   * @param eventName
```
Comment (thiagowfx) [inline src/bidiMapper/domains/events/EventManager.ts:76]: nit: value maps

File: src/bidiMapper/domains/events/EventManager.ts [hunk 3]
```diff
@@ -47,30 +71,64 @@ export class EventManager implements IEventManager {
...
     contextId: CommonDataTypes.BrowsingContext | null
   ): Promise<void> {
+    const eventWrapper = new IdWrapper<Message.EventMessage>(event);
     const sortedChannels =
       this.#subscriptionManager.getChannelsSubscribedToEvent(
         event.method,
         contextId
       );
-
+    // Buffers event if needed.
```
Comment (thiagowfx) [inline src/bidiMapper/domains/events/EventManager.ts:99]: This comment seems redundant to me, I'd remove it.

File: src/bidiMapper/domains/events/EventManager.ts [hunk 4]
```diff
@@ -86,4 +144,89 @@ export class EventManager implements IEventManager {
...
+    const bufferMapKey = EventManager.#getMapKey(eventName, contextId);
+    const lastSentMapKey = EventManager.#getMapKey(
+      eventName,
+      contextId,
+      channel
+    );
+    const lastSentMessageId =
+      this.#lastMessageSent.get(lastSentMapKey) ?? -Infinity;
+
+    return (
```
Comment (thiagowfx) [inline src/bidiMapper/domains/events/EventManager.ts:225]: redundant return parens

**Review by thiagowfx — COMMENTED** [review]
The data structures look great!

File: src/bidiMapper/domains/events/EventManager.ts (hunk 4 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts:225]: agree )

File: src/bidiMapper/domains/events/EventManager.ts (hunk 3 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts:99]: agree )

File: src/bidiMapper/domains/events/EventManager.ts (hunk 4 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts:225]: It's `prettier`

File: src/bidiMapper/domains/events/EventManager.ts (hunk 2 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts:76]: thanks!

File: src/bidiMapper/domains/events/EventManager.ts (hunk 1 above)
Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts:209]: It is auto-generated by IntelliJ.

File: src/utils/buffer.spec.ts [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import * as chai from 'chai';
+import { Buffer } from './buffer';
+
+import * as sinon from 'sinon';
+
+const expect = chai.expect;
+
+describe('test Buffer', () => {
+  it('should keep values', () => {
+    const buffer = new Buffer<number>(2);
```
[the remark this answers is not in the document]
Comment (sadym-chromium) [reply src/utils/buffer.spec.ts:27]: done. Thanks!

File: src/bidiMapper/domains/events/EventManager.ts (hunk 1 above)
Comment (thiagowfx) [reply src/bidiMapper/domains/events/EventManager.ts:238]: Then we should remove them. In fact, we can even remove the `@param`s. It's redundant jsdoc noise.

Comment (sadym-chromium) [reply src/bidiMapper/domains/events/EventManager.ts]: Removed in this file, but I'd reduce the amount of manual corrections to as low as possible.

**Review by thiagowfx — APPROVED** [review]

**Review by thiagowfx — APPROVED** [review]

**Pull request merged**